dlib/matrix/matrix_fft.h

```diff
@@ -9,837 +9,142 @@
-            << "\n\t is_power_of_two(data.nr()): " << is_power_of_two(data.nr())
-            << "\n\t is_power_of_two(data.nc()): " << is_power_of_two(data.nc())
-            );
-
-        if (data.size() == 0)
-            return;
-
-        DFTI_DESCRIPTOR_HANDLE h;
-        MKL_LONG status;
-
-        if (data.nr() == 1 || data.nc() == 1)
-        {
-            status = DftiCreateDescriptor(&h, DFTI_DOUBLE, DFTI_COMPLEX, 1, data.size());
-            DLIB_DFTI_CHECK_STATUS(status);
-        }
-        else
-        {
-            MKL_LONG size[2];
-            size[0] = data.nr();
-            size[1] = data.nc();
-
-            status = DftiCreateDescriptor(&h, DFTI_DOUBLE, DFTI_COMPLEX, 2, size);
-            DLIB_DFTI_CHECK_STATUS(status);
-
-            MKL_LONG strides[3];
-            strides[0] = 0;
-            strides[1] = size[1];
-            strides[2] = 1;
-
-            status = DftiSetValue(h, DFTI_INPUT_STRIDES, strides);
-            DLIB_DFTI_CHECK_STATUS(status);
-        }
-
-        // Unless we use sequential mode, the fft results are not correct.
-        status = DftiSetValue(h, DFTI_THREAD_LIMIT, 1);
-        DLIB_DFTI_CHECK_STATUS(status);
-
-        status = DftiCommitDescriptor(h);
-        DLIB_DFTI_CHECK_STATUS(status);
-
-        if (do_backward_fft)
-            status = DftiComputeBackward(h, &data(0, 0));
-        else
-            status = DftiComputeForward(h, &data(0, 0));
-        DLIB_DFTI_CHECK_STATUS(status);
-
-        status = DftiFreeDescriptor(&h);
-        DLIB_DFTI_CHECK_STATUS(status);
-
-        return;
+        kiss_fft({(int)data.nr(), (int)data.nc()}, &data(0,0), &data(0,0), false);
     }
 
 // ----------------------------------------------------------------------------------------
 
-    // Call the MKL DFTI implementation in these cases 
-
-    inline matrix<std::complex<double>,0,1> fft (const matrix<std::complex<double>,0,1>& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double>,0,1> ifft(const matrix<std::complex<double>,0,1>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double>,1,0> fft (const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double>,1,0> ifft(const matrix<std::complex<double>,1,0>& data)
-    {
-        return call_mkl_fft(data, true) / data.size();
-    }
-    inline matrix<std::complex<double> > fft (const matrix<std::complex<double> >& data)
-    {
-        return call_mkl_fft(data, false);
-    }
-    inline matrix<std::complex<double> > ifft(const matrix<std::complex<double> >& data)
+    template < typename T, long NR, long NC, typename MM, typename L >
+    void ifft_inplace (matrix<std::complex<T>,NR,NC,MM,L>& data)
     {
-        return call_mkl_fft(data, true) / data.size();
+        kiss_fft({(int)data.nr(), (int)data.nc()}, &data(0,0), &data(0,0), true);
```

Comment:
Make the type of the dims argument `long` so that these `(int)` casts are not required.